Service/MysqlDumpService.php

```diff
@@ -0,0 +1,176 @@
+     * @return array
+     *
+     * @throws DirectoryCreationException
+     * @throws InvalidConnectionException
+     * @throws InvalidConnectionTypeException
+     * @throws NullConnectionException
+     * @throws BackupDeletionException
+     */
+    public function dump (DatabaseConnection $connection)
+    {
+        if (is_null($connection))
+        {
+            throw new NullConnectionException("The connection can't be null.");
+        }
+
+        if ($connection->getType() !== DatabaseConnection::TYPE_MYSQL)
+        {
+            throw new InvalidConnectionTypeException("Can't perform backup on non-MySQL Databases.");
+        }
+
+        if (!$this->validateConnection($connection))
+        {
+            throw new InvalidConnectionException("Could not establish connection to the given connection '{$connection->getIdentifier()}'.");
+        }
+
+        if (!$this->createBackupDirectory($connection->getBackupPath()))
+        {
+            throw new DirectoryCreationException("Permission denied. Could not create directory {$connection->getBackupPath()}");
+        }
+
+        $process = new Process(sprintf(
+            'mysqldump --user="%s" --password="%s" --host="%s" --lock-all-tables "%s" | gzip > "%s"',
+            $connection->getUsername(),
+            $connection->getPassword(),
+            $connection->getHost(),
+            $connection->getDatabase(),
+            $connection->getBackupPath()
+        ));
+        $process->run();
+
+
+        // We need to determine ourselves whether mysqldump has raised an error
+        // as its return code is unreliable due to the fact that it's always returning 0
+        $errorOut = trim($process->getErrorOutput());
+        $success  = (preg_match('/Got error:.*/', $errorOut) !== 1);
+
+        if (!$success && !$this->removeFaultyBackup($connection->getBackupPath()))
+        {
+            throw new BackupDeletionException(
+                "Could not remove faulty backup of database '{$connection->getDatabase()}' (connection: {$connection->getIdentifier()}) at {$connection->getBackupPath()}.");
+        }
+
+        return [
+            'success' => $success,
+            'error'   => $errorOut
+        ];
+    }
+
+
+    /**
+     * Determines whether the given connection is valid
+     *
+     * @param DatabaseConnection $connection
+     *
+     * @return bool
+     */
+    public function validateConnection (DatabaseConnection $connection)
+    {
+        try
+        {
+            $driver = new mysqli_driver();
+            // Backup the current configuration to prevent any side effects our reporting mode may cause
+            $originalReportMode = $driver->report_mode;
+
+            // Ensure that MySQLi is throwing proper exceptions
+            $driver->report_mode = MYSQLI_REPORT_ALL;
+
+            $mysqli = new mysqli($connection->getHost(), $connection->getUsername(), $connection->getPassword(), $connection->getDatabase(), $connection->getPort());
```

Comment:
My impression was that we would simply create dedicated *DumpService classes for each database type we add in the future anyway, which could also handle the connection validation differently. 

I don't mind which way we implement it as long as we decide on one :)